tests/integration-tests/tests/integration/catalog_tests.rs

```diff
@@ -636,3 +642,187 @@ async fn test_multiple_cname_additionals() {
+
+        let mut catalog = Catalog::new();
+        catalog.upsert(origin.into(), vec![Arc::new(records)]);
+        catalog
+    }
+
+    async fn run_query(catalog: &Catalog, query: Query) -> Message {
+        let mut question = Message::new();
+        question.add_query(query);
+        question
+            .extensions_mut()
+            .get_or_insert_with(Edns::new)
+            .enable_dnssec();
+
+        let question_bytes = question.to_bytes().unwrap();
+        let question_req = MessageRequest::from_bytes(&question_bytes).unwrap();
+        let question_req = Request::new(question_req, ([127, 0, 0, 1], 5553).into(), Protocol::Udp);
+
+        let response_handler = TestResponseHandler::new();
+        catalog
+            .lookup(&question_req, None, response_handler.clone())
+            .await;
+        response_handler.into_message().await
+    }
+
+    #[tokio::test]
+    async fn test_dnskey_and_nsec3() {
+        let catalog = make_catalog();
+
+        let mut query = Query::new();
+        query.set_name(Name::from_str("test.com.").unwrap());
+        query.set_query_type(RecordType::DNSKEY);
+
+        // Check DNSKEY + RRSIG
+        {
+            let result = run_query(&catalog, query).await;
+
+            let dnskey = result
+                .answers()
+                .iter()
+                .find(|e| e.record_type() == RecordType::DNSKEY)
+                .expect("result to contain one DNSKEY");
+            let rrsig = result
+                .answers()
+                .iter()
+                .find(|e| e.record_type() == RecordType::RRSIG)
+                .expect("result to contain one DNSKEY");
+            assert_eq!(result.answers().len(), 2, "expect only one answer");
+
+            let dnskey = dnskey
+                .data()
+                .clone()
+                .into_dnssec()
+                .unwrap()
+                .into_dnskey()
+                .unwrap();
+            assert!(dnskey.zone_key());
+
+            let rrsig = rrsig
+                .data()
+                .clone()
+                .into_dnssec()
+                .unwrap()
+                .into_rrsig()
+                .unwrap();
+            assert_eq!(rrsig.type_covered(), RecordType::DNSKEY);
+        }
+
+        // Check NSEC3
+        {
+            let mut query = Query::new();
+            query.set_name(Name::from_str("test.com.").unwrap());
+            query.set_query_type(RecordType::NSEC);
+
+            let result = run_query(&catalog, query).await;
+            assert!(result.answers().is_empty());
+
+            result
+                .name_servers()
+                .iter()
+                .find(|e| e.record_type() == RecordType::SOA)
+                .expect("authority section to contains SOA");
+
+            let nsec3 = result
+                .name_servers()
+                .iter()
+                .find(|e| e.record_type() == RecordType::NSEC3)
+                .expect("result to contain NSEC");
```

Comment:
nit: incorrect error message
```suggestion
                .expect("result to contain NSEC3");
```